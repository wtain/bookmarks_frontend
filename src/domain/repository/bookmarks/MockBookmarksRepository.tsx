
import React from "react";
import BookmarkDto from "../../dto/BookmarkDto";
import IBookmarksRepository from "./IBookmarksRepository";
import ReactHelpers from "../../../utils/ReactHelpers"
import BookmarkUtils from "../../../utils/BookmarkUtils";

class MockBookmarksRepository implements IBookmarksRepository {

    private bookmarks: BookmarkDto[];

    constructor() {
        this.bookmarks = Array.from(Array(20).keys()).map(i => 
            BookmarkUtils.createBookmark("Summary " + i, 
                "Contents " + i + "\nLorem ipsum\nLorem ipsum", 
                Array.from(Array(i).keys()).map(j => { 
                    return { ...BookmarkUtils.createNewTag(), name: "tag " + j }
                })
            ));
    }

    async getBookmark(id: string): Promise<BookmarkDto> {
        // todo: implement
        await ReactHelpers.delay(300);
        return this.bookmarks[0];
    }

    public async getBookmarks(): Promise<BookmarkDto[]> {
        await ReactHelpers.delay(300);
        return this.bookmarks;
    }

    async getBookmarksByTag(tag: string): Promise<BookmarkDto[]> {
        await ReactHelpers.delay(300);
        return this.bookmarks; // todo: implement
    }

    public async addBookmark(bookmark: BookmarkDto) {
        this.bookmarks = [...this.bookmarks, bookmark]
    }

    public async removeBookmark(bookmark: BookmarkDto) {
        this.bookmarks = this.bookmarks.filter(b => b.id != bookmark.id);
    }

    public async editBookmark(bookmark: BookmarkDto) {
        this.bookmarks = this.bookmarks.map(b => b.id === bookmark.id ? bookmark : b);
    }
}

export default MockBookmarksRepository;