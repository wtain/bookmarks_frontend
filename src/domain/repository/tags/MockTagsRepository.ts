import ReactHelpers from '../../../utils/ReactHelpers';
import ITagsRepository from './ITagsRepository';


class MockTagsRepository implements ITagsRepository {
  async getTags(): Promise<string[]> {
    await ReactHelpers.delay(300);
    return ["tag1", "tag2", "tag3"];
  }
  
}

export default MockTagsRepository;