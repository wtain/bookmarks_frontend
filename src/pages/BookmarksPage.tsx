
import useEventListener from "@use-it/event-listener";
import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import BookmarkForm from "../components/UI/BookmarkForm/BookmarkForm";
import BookmarkList from "../components/UI/BookmarkList/BookmarkList";
import Loading from "../components/UI/Loading/Loading";
import Modal from "../components/UI/Modal/Modal";
import BookmarkDto from "../domain/dto/BookmarkDto";
import TagDto from "../domain/dto/TagDto";
import IBookmarksRepository from "../domain/repository/bookmarks/IBookmarksRepository";
import cl from './BookmarksPage.module.css'

interface Props {
    bookmarksRepository: IBookmarksRepository;
}

const BookmarksPage = (props: Props) => {

    const [modal, setModal] = useState(false);

    const { tag, id } = useParams();

    let [newBookmarkId, setNewBookmarkId] = useState("");

    let [bookmarks, setBookmarks] = useState(
        [] as BookmarkDto[]
    )

    let [loading, setLoading] = useState(false);

    const getData = async (): Promise<BookmarkDto[]> => {
        if (id) {
            return [await props.bookmarksRepository.getBookmark(id)];
        }
        return await (tag ? 
                props.bookmarksRepository.getBookmarksByTag(tag) : 
                props.bookmarksRepository.getBookmarks());
    }

    const loadBookmarks = async (success: (bookmarks: BookmarkDto[]) => void, error: (e: any) => void) => {
        try {
            const bookmarks = await getData();
            success(bookmarks)
        } catch (e) {
            error(e)
        }
    }

    useEffect(() => {
        setLoading(true);
        loadBookmarks((bookmarks: BookmarkDto[]) => {
            setBookmarks(bookmarks)
            setLoading(false);
        }, 
            (e) => {
                setLoading(false);
            })
    }, [tag, id])

    const doUpdatePoll = async () => {
        setBookmarks(
            await getData()
            );
    }

    const addBookmark = async (newBookmark: BookmarkDto) => {
        setModal(false);
        setNewBookmarkId(newBookmark.id);
        await props.bookmarksRepository.addBookmark(newBookmark);
        doUpdatePoll();
    }

    const removeBookmark = async (bm: BookmarkDto) => {
        setNewBookmarkId("");
        await props.bookmarksRepository.removeBookmark(bm);
        doUpdatePoll();
    }

    useEffect(() => {
        const interval = setInterval(() => doUpdatePoll(), 20 * 1000);
        return () => clearInterval(interval);
    }, [tag, id]);

    useEventListener('keydown', (e: KeyboardEvent) => {
        if (e.key === "a" && e.ctrlKey) {
            setModal(true);
            e.stopPropagation();
        } else if (e.key === "Escape" && modal) {
            setModal(false);
        }
    })

    return (
        <div>
            <Modal visible={modal} setVisible={setModal}>
                <BookmarkForm 
                        create={addBookmark} 
                        visibility={modal}
                        validation={(bookmark: BookmarkDto) => bookmark.contents.trim().length > 0 && bookmark.summary.trim().length > 0} />
            </Modal>
            <button onClick={() => setModal(true)} className={cl.btn_add}>
                Add...
            </button>

            {
                loading ? 
                <Loading /> 
                : 
                <BookmarkList bookmarks={bookmarks} 
                        newBookmarkId={newBookmarkId}
                        onRemoveBookmark={(bm: BookmarkDto) => removeBookmark(bm)} 
                        onBookmarkContentsChanged={async (bm: BookmarkDto, new_contents: string) => {
                            await props.bookmarksRepository.editBookmark({...bm, contents: new_contents, updated: new Date().toLocaleString()});
                            doUpdatePoll();
                        }}
                        onBookmarkSummaryChanged={async (bm: BookmarkDto, new_summary: string) => {
                            await props.bookmarksRepository.editBookmark({...bm, summary: new_summary, updated: new Date().toLocaleString()});
                            doUpdatePoll();
                        }}
                        onBookmarkTagAdded={async (bm: BookmarkDto, new_tag: TagDto) => {
                            await props.bookmarksRepository.editBookmark({...bm, tags: [...bm.tags, new_tag], updated: new Date().toLocaleString()});
                            doUpdatePoll();
                        }}
                        onBookmarkTagRemoved={async (bm: BookmarkDto, index: number) => {
                            await props.bookmarksRepository.editBookmark({...bm, tags: bm.tags.filter((v, i) => i !== index), updated: new Date().toLocaleString()});
                            doUpdatePoll();
                        }}
                        onBookmarkIsDoneChanged={async (bm: BookmarkDto, new_value: boolean) => {
                            await props.bookmarksRepository.editBookmark({...bm, isDone: new_value, updated: new Date().toLocaleString()});
                            doUpdatePoll();
                        }}
                        />
            }
            
        </div>
    )
}

export default BookmarksPage;