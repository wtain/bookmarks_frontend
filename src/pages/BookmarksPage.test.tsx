import renderer from 'react-test-renderer';
import BookmarksPage from './BookmarksPage';
import MockBookmarksRepository from '../domain/repository/bookmarks/MockBookmarksRepository';
import MockTagsRepository from '../domain/repository/tags/MockTagsRepository';
import { BrowserRouter } from 'react-router-dom';

// test('renders learn react link', () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });

it('BookmarksPage renders correctly', () => {
  const bookmarksRepository = new MockBookmarksRepository();
  const tagsRepository = new MockTagsRepository();

  const tree = renderer
    .create(
      <BrowserRouter>
        <BookmarksPage bookmarksRepository={bookmarksRepository} tagsRepository={tagsRepository} />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});