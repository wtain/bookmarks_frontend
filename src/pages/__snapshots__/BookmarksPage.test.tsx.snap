// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarksPage renders correctly 1`] = `
<div>
  <div
    className="Modal"
    onClick={[Function]}
  >
    <div
      className="ModalContent"
      onClick={[Function]}
    >
      <form
        className="bookmark_form form_invalid"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Summary"
          value=""
        />
        <div
          className="textarea_wrapper"
        >
          <textarea
            className="textarea"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Contents"
            rows={3}
            value=""
          />
        </div>
        <div
          className="container tags_container"
          onClick={[Function]}
        >
          <span
            className="sp"
          >
            <input
              className="input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Tags"
              value=""
            />
          </span>
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
          Add
        </button>
      </form>
    </div>
  </div>
  <button
    className="btn_add"
    onClick={[Function]}
  >
    Add...
  </button>
  <span
    className="tag"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    (Clear)
  </span>
  <div>
    (No bookmarks)
  </div>
</div>
`;

exports[`BookmarksPagerenders correctly 1`] = `
<div>
  <div
    className="Modal"
    onClick={[Function]}
  >
    <div
      className="ModalContent"
      onClick={[Function]}
    >
      <form
        className="bookmark_form form_invalid"
      >
        <input
          className="input"
          onChange={[Function]}
          placeholder="Summary"
          value=""
        />
        <div
          className="textarea_wrapper"
        >
          <textarea
            className="textarea"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Contents"
            rows={3}
            value=""
          />
        </div>
        <div
          className="container tags_container"
          onClick={[Function]}
        >
          <span
            className="sp"
          >
            <input
              className="input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Tags"
              value=""
            />
          </span>
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
          Add
        </button>
      </form>
    </div>
  </div>
  <button
    className="btn_add"
    onClick={[Function]}
  >
    Add...
  </button>
  <span
    className="tag"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    (Clear)
  </span>
  <div>
    (No bookmarks)
  </div>
</div>
`;
